projects/components/src/file-upload/file-upload.service.ts

```diff
@@ -0,0 +1,127 @@
+import { HttpClient, HttpEventType } from '@angular/common/http';
+import { Injectable } from '@angular/core';
+import { uniqueId } from 'lodash-es';
+import { fromEvent, Observable, of, Subscription } from 'rxjs';
+import { catchError, delay, filter, finalize, map, shareReplay } from 'rxjs/operators';
+import { NotificationService } from '../notification/notification.service';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class FileUploadService {
+  private readonly activeSubscriptionsMap: Map<string, Subscription> = new Map<string, Subscription>();
+  private beforeUnloadSubscription?: Subscription;
+
+  public constructor(private readonly http: HttpClient, private readonly notificationService: NotificationService) {}
+
+  public uploadFilesAsFormDataAsync<TResponse extends object>(
+    url: string,
+    formData: FormData
+  ): Observable<FileUploadEvent<TResponse>> {
+    if (this.beforeUnloadSubscription === undefined) {
+      this.addBeforeUnloadHandler();
+    }
+
+    const fileUploadId = uniqueId('file-upload');
+    const upload$ = this.uploadFilesAsFormData<TResponse>(url, formData).pipe(
+      delay(100),
+      finalize(() => {
+        this.onUploadComplete(fileUploadId);
+      }),
```

Comment:
do we expect failures here?

Reply:
Yes, it could fail. We are capturing error in `uploadFilesAsFormData`

Reply:
finalize would be called on both success and failure. is that our intention?

if rest of the logic is in the `uploadFilesAsFormData` method, it'd make sense for this also to be moved there so that all the operators are in one place and readable. Or lift up the success/failure handling here. Both works.

Reply:
This method just does a long lived subscription. I am not sure i understand the change